src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs

```diff
@@ -0,0 +1,489 @@
+
+                        // Write the chunk.
+                        this.WriteChunkWithoutEscaping(combinedArray, totalLength, isFinalBlock);
+
+                        ArrayPool<char>.Shared.Return(combinedArray);
+                    }
+                    else
+                    {
+                        this.WriteChunkWithoutEscaping(chunk.Span, chunk.Length, isFinalBlock);
+                    }
+
+                    // Flush the writer if the buffer threshold is reached
+                    await this.jsonWriter.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+                }
+            }
+
+            private void WriteChunk(ReadOnlySpan<char> chunk, int chunkLength, int firstIndexToEscape, bool isFinalBlock)
+            {
+                if (firstIndexToEscape != -1)
+                {
+                    this.jsonWriter.WriteEscapedStringChunk(chunk, firstIndexToEscape, isFinalBlock, out numOfCharsNotWrittenFromPreviousChunk);
+
+                    if (this.numOfCharsNotWrittenFromPreviousChunk > 0)
+                    {
+                        if (this.buffer == null)
+                        {
+                            this.buffer = ArrayPool<char>.Shared.Rent(ODataUtf8JsonWriter.chunkSize);
+                        }
+
+                        // Update the buffer with unprocessed bytes from the current chunk.
+                        chunk.Slice(chunkLength - this.numOfCharsNotWrittenFromPreviousChunk).CopyTo(this.buffer.AsSpan());
+                    }
+                }
+                else
+                {
+                    this.jsonWriter.WriteStringChunk(chunk, isFinalBlock);
+                }
+            }
+
+            private void WriteStringChunkWithoutEscaping(ReadOnlySpan<char> chunk, bool isFinalBlock, out int charsFromPrevChunkNotProcessed)
+            {
+                int maxUtf8Length = Encoding.UTF8.GetMaxByteCount(chunk.Length);
+
+                Span<byte> output = this.jsonWriter.bufferWriter.GetSpan(maxUtf8Length);
+                OperationStatus status = Utf8.FromUtf16(chunk, output, out int charsRead, out int charsWritten, isFinalBlock);
+                Debug.Assert(status == OperationStatus.Done || status == OperationStatus.NeedMoreData);
+
+                charsFromPrevChunkNotProcessed = chunk.Length - charsRead;
+
+                // notify the bufferWriter of the write
+                this.jsonWriter.bufferWriter.Advance(charsWritten);
+            }
+
+            private void WriteChunkWithoutEscaping(ReadOnlySpan<char> chunk, int chunkLength, bool isFinalBlock)
+            {
+                this.WriteStringChunkWithoutEscaping(chunk, isFinalBlock, out numOfCharsNotWrittenFromPreviousChunk);
+
+                if (this.numOfCharsNotWrittenFromPreviousChunk > 0)
+                {
+                    if (this.buffer == null)
+                    {
+                        this.buffer = ArrayPool<char>.Shared.Rent(ODataUtf8JsonWriter.chunkSize);

```

Comment:
Why not:
```csharp
this.buffer = ArrayPool<byte>.Shared.Rent(this.numBytesNotWrittenFromPreviousChunk);
```

Reply:
So we rent the buffer once and use it for the entire stream scope. If in the first chunk the numOfChars not processed is less than the numOfCharsNotProcessed in the next chunk.. This will lead to problems ..

Reply:
I don't mean to flog a dead horse but would you consider a constant that is big enough for unprocessed bytes but not as big as 2048 set as chunk size? From what I understood from @habbes review comments, the unprocessed bytes will be a maximum 3 or thereabouts. If we're trying to manage COGS, 2048 feels like an overkill for that. You could set a class constant specific for that instead of reusing a constant that was intended for an entirely different thing and in the process clawing back on the COGS gains.
@habbes Did I understand the maximum number of unprocessed bytes correctly?

Reply:
It was hard to determine the right buffer size to use for the unprocessed bytes. Sometimes the entire chunk might not processed. Then what happens at that time? I tried different buffer sizes and I kept getting NeedMoreData. 10, 20, 100. This doesn't always happen but it can happen. I'm not sure whether this makes sense.